doc/kernel/drivers/index.rst

```diff
@@ -122,36 +125,48 @@ Most drivers will be implementing a device-independent subsystem API.
 Applications can simply program to that generic API, and application
 code is not specific to any particular driver implementation.
 
+If all driver API instances are assigned to their respective API linker section
+use :c:macro:`DEVICE_API_IS()` to verify the API's type.
+
 A subsystem API definition typically looks like this:
 
 .. code-block:: C
 
   typedef int (*subsystem_do_this_t)(const struct device *dev, int foo, int bar);
   typedef void (*subsystem_do_that_t)(const struct device *dev, void *baz);
 
-  struct subsystem_api {
+  __subsystem struct subsystem_driver_api {
         subsystem_do_this_t do_this;
         subsystem_do_that_t do_that;
   };
 
   static inline int subsystem_do_this(const struct device *dev, int foo, int bar)
   {
-        struct subsystem_api *api;
+        struct subsystem_driver_api *api;
+
+        api = (struct subsystem_driver_api *)dev->api;
+        CHECKIF(!DEVICE_API_IS(subsystem, api)) {
+              return -ENODEV;
+        }
```

Comment:
`CHECKIF` can assert, but it is up to the desired configuration.

https://github.com/zephyrproject-rtos/zephyr/blob/e141640b8259c337f24c481c1a354c072556a996/include/zephyr/sys/check.h#L13-L23

Reply:
even if possible, it doesn't make sense to propagate errors that can't be handled. It's not an unexpected failure, but an application bug if this situation is reached.

Reply:
I would argue that it could be valid for runtime calls on devices that are retrieved using `device_get_binding`. Also, I think runtime checks are enforced when using `CONFIG_USERSPACE`, but I could be wrong.

Callers would typically check the return value for API calls anyway.

Reply:
I buy the device_get_binding case (not that common nowadays, but used in eg shells). Let's keep as is then.